pymtl3/stdlib/ifcs/CalleeIfc.py

```diff
@@ -0,0 +1,167 @@
+#=========================================================================
+# CalleeIfc.py
+#=========================================================================
+# RTL method interface. RTL equivalence of NonBlockingCalleeIfc in CL
+#
+# Author : Yixiao Zhang
+#   Date : June 10, 2019
+
+from __future__ import absolute_import, division, print_function
+
+from copy import deepcopy
+
+from pymtl3 import *
+from pymtl3.dsl.errors import InvalidConnectionError
+
+
+#-------------------------------------------------------------------------
+# CalleeRTL2CL
+#-------------------------------------------------------------------------
+class CalleeRTL2CL( Component ):
+
+  def construct( s, ifc_rtl_callee ):
+
+    ifc_rtl_caller = deepcopy( ifc_rtl_callee )
+    ifc_rtl_caller._dsl.constructed = False
+    s.ifc_rtl_caller = ifc_rtl_caller.inverse()
+
+    s.called = False
+    s.rdy = False
+
+    # generate upblk depending on args
+    if s.ifc_rtl_caller.args:
+      # has args, create tmp var
+      s.args = s.ifc_rtl_caller.args._dsl.Type()
+
+      # update args & en
+      @s.update
+      def update_en_args():
+        s.ifc_rtl_caller.en = Bits1( 1 ) if s.called else Bits1( 0 )
+        s.ifc_rtl_caller.args = s.args
+        s.called = False
+
+      # add constraints between callee method and upblk
+      s.add_constraints( M( s.cl_callee_method ) < U( update_en_args ) )
+
+      # know which method to add constraints on later
+      cl_method = s.cl_callee_method
+
+    else:
+      # no args, update en
+      @s.update
+      def update_en():
+        s.ifc_rtl_caller.en = Bits1( 1 ) if s.called else Bits1( 0 )
+        s.called = False
+
+      # add constraints between callee method and upblk
+      s.add_constraints( M( s.cl_callee_method_no_arg ) < U( update_en ) )
+
+      # know which method to add constraints on later
+      cl_method = s.cl_callee_method_no_arg
+
+    # generate upblk depending on rets
+    if s.ifc_rtl_caller.rets:
+      # create tmp var for rets
+      s.rets = s.ifc_rtl_caller.rets._dsl.Type()
+
+      # generate upblk for rets
+      @s.update
+      def update_rets():
+        s.rets = s.ifc_rtl_caller.rets
+
+      s.add_constraints( U( update_rets ) < M( cl_method ) )
+
+    else:
+      # return None if no ret
+      s.rets = None
+
+    # Generate upblk and add constraints for rdy
+    @s.update
+    def update_rdy():
+      s.rdy = True if s.ifc_rtl_caller.rdy else False
+
+    s.add_constraints( U( update_rdy ) < M( cl_method ) )
+
+  @non_blocking( lambda s: s.rdy )
+  def cl_callee_method( s, **kwargs ):
```

Comment:
I think a cleaner way to do so, is to have a separate ready method and two methods. You can instantiate an empty NonBlockingCalleeIfc() in construct, and based on the parameters, you set the calleeifc's method to either cl_callee_method, or cl_callee_method_no_args.

Reply:
Putting non_blocking twice will create two bundles before you enter the construct

Reply:
fixed -> https://github.com/cornell-brg/pymtl3/blob/6db73d99755ce8c11ed2b97b30751cadec0f76c2/pymtl3/stdlib/ifcs/generic_method_ifcs.py#L46-L59